refactor: change ConfigWriter interface name to Configurator

sharikovvladislav opened the pull request:
Try to fix https://github.com/stryker-mutator/stryker/issues/311

TODO: 
- [x] review issues
- [x] add clean script to the root `package.json`: `"clean": "lerna run prebuild"`
- [x] fix prebuild (perhaps add clean script) in subpackages to clean `reports` dirs (like @nicojs sugessted [there](https://github.com/stryker-mutator/stryker/pull/357#issuecomment-322444701))

sharikovvladislav commented:
ok! I need some time for this

sharikovvladislav commented:
Is there any way to delete generated files? I ran `npm test` and now I have a lot of generated files. It became a bit harder to track all file names, since they also used in generated files.

nicojs commented:
In the sub package directory you should me able to use `npm run clean`. However, you might also be able to just hide all `*.d.ts`, `*.js` and`*.map` files in your editor. Vscode is already configures for this, uwing the .vscode hidden folder we check in

sharikovvladislav commented:
I have this at the moment: http://i.imgur.com/u3aWvbX.png

So I can't see *.js in the side bar, but I can select it through "Find file" field (CMD + P). Is this planned?

Why I am talking about this. I want to rename file and I have to refactor imports. I want to find all imports of ConfigWriter. Here is the result: http://i.imgur.com/RE3axXx.png. As you see, there is also `*.ts.html` file. It is some report (I think it was generated by `npm test`). Thats why I ask about some command to clean all reports. By the way: can I somehow run `npm run clean` in the root and clean everything in subpackages?

Also another question: is there any plugin in VSCode to automatically refactor imports (like WebStorm do) when I refactor file name?

nicojs commented:
> By the way: can I somehow run npm run clean in the root and clean everything in subpackages?

Hmm i see that the script `"clean"` is not implemented in every package. Some packages use the `"prebuilt"` script directly. If you add this script to the root package.json: `"clean": "lerna run prebuild",` you should be able to run `npm run clean` from the root package.

> Also another question: is there any plugin in VSCode to automatically refactor imports (like WebStorm do) when I refactor file name?

Some refactoring is done automatically if you press f2 on ConfigWriter. However, since we're using a lerna mono repo, vscode does not know that the `import { ConfigWriter } from 'stryker-api/config';` in "stryker\packages\stryker-karma-runner\src\KarmaConfigWriter.ts" points to `export {default as ConfigWriter} from './src/config/ConfigWriter';` in "stryker\packages\stryker-karma-runner\src\KarmaConfigWriter.ts" which in turn points to `export default ConfigWriter;` in "stryker\packages\stryker-api\src\config\ConfigWriter.ts"

Not sure if there is a better way to rename. Find and replace might be the best strategy here.

sharikovvladislav commented:
Its ok about F2 and renaming class/interface name. Yes everything except `import { ConfigWriter } from 'stryker-api/config';` will be renamed. Its ok.

I am asking about refactoring file name. If you refactor file name in WebStorm all imports will be refactored too. I just asking if vscode has some plugin for what. Ok I got it. Ty!

nicojs commented:
> I am asking about refactoring file name. If you refactor file name in WebStorm all imports will be refactored too. I just asking if vscode has some plugin for what.

This seems to be a good plugin for renaming: https://marketplace.visualstudio.com/items?itemName=stringham.move-ts. But again, not sure if it works with the lerna mono repo setup we have

> Should I add clean in the main package.json in that PR?

Yes, i think its a good addition to our build tools.

sharikovvladislav commented:
1) I tried TS Move. Ye all imports was refactored too, but associated *.d.ts file was not renamed. WS doing this rename too (it suggests and I have to approve it) :(
2) `"clean": "lerna run prebuild"`. Looks like clean helped a bit, but there are still many *reports* dirs with *.ts.html files.

nicojs commented:
> I tried TS Move. Ye all imports was refactored too, but associated *.d.ts file was not renamed. WS doing this rename too (it suggests and I have to approve it) :(
Just to be clear, you don't need to refactor the *.d.ts files. Those are generated by TypeScript.

> `"clean": "lerna run prebuild"`. Looks like clean helped a bit, but there are still many reports dirs with *.ts.html files.

It's always save to remove `reports` directories as it is build output. Apparently we still need to add the removal of that folder to all clean tasks. Don't worry about it.

sharikovvladislav commented:
> Don't worry about it.

Ok, I will write small temporary script for myself. Should I add this to this PR? Or you will do it by yourself?

nicojs commented:
> Ok, I will write small temporary script for myself. Should I add this to this PR? Or you will do it by yourself?

If you really want to commit it, i would suggest add it to the prebuild scripts of the different packages.

For example: 

`""prebuild": "rimraf \"+(*.js|*.map)\" \"src/**/+(*.js|*.map)\" \"test/**/+(*.js|*.map)\" \"testResources/module/+(*.js|*.map)\" \"testResources/module/node_modules/stryker-api\"`

Becomes

`""prebuild": "rimraf \"+(*.js|*.map)\" \"src/**/+(*.js|*.map)\" \"test/**/+(*.js|*.map)\" \"testResources/module/+(*.js|*.map)\" \"testResources/module/node_modules/stryker-api\" \"reports\"`

sharikovvladislav commented:
OK. I added todo list in the first message. Now I need sometime to understand this and do some refactoring :)

nicojs commented:
Awesome, if you need more help you can always ask us here or on gitter.

sharikovvladislav commented:
If everything will be ok with that PR I will have questions about 2nd issue where I also left the comment. But lets do it later.

sharikovvladislav commented:
I added clean script to the root package.json and also I fixed prebuild scripts in packages (I created clean script and I call it from prebuild). What do you think?

nicojs commented on packages/stryker-api/package.json:
```diff
@@ -4,7 +4,8 @@
   "scripts": {
     "start": "tsc -w",
-    "prebuild": "rimraf \"+(*.js|*.map)\" \"src/**/+(*.js|*.map)\" \"test/**/+(*.js|*.map)\" \"testResources/module/+(*.js|*.map)\" \"testResources/module/node_modules/stryker-api\"",
+    "clean": "rimraf \"+(*.d.ts|*.js|*.map)\" \"src/**/+(*.d.ts|*.js|*.map)\" \"test/**/+(.d.ts|*.js|*.map)\" \"testResources/module/+(.d.ts|*.js|*.map)\" \"testResources/module/node_modules/stryker-api\" reports",
```
Shouldn't it be  `"clean": "rimraf \"+(*.d.ts|*.js|*.map)\" \"src/**/+(*.d.ts|*.js|*.map)\" \"test/**/+(*.d.ts|*.js|*.map)\" \"testResources/module/+(*.d.ts|*.js|*.map)\" \"testResources/module/node_modules/stryker-api\" reports"` ?

sharikovvladislav commented:
Cool! I did all what you said in the review. I also found some extra issues. I fixed everything.

What do you think?

sharikovvladislav commented:
By the way, I found some directory named `.nyc_output`. I think it is not bad idea to rimraf this dir too in clean task. What do you think?

nicojs commented on packages/stryker-api/src/config/ConfigEditor.ts:
```diff
@@ -1,7 +1,7 @@
 
-interface ConfigWriter {
+interface ConfigEditor {
   write(config: Config): void;
```
We should also rename `write` to `edit`.

simondel commented on packages/stryker-api/src/config/ConfigWriter.ts:
```diff
@@ -1,7 +1,7 @@
 import Config from './Config';
```
This has been fixed

sharikovvladislav commented:
> We should also rename write to edit.

Done. I got red tests again, but I think it is not because of my change :( Lets see what CI say.

 > By the way, I found some directory named .nyc_output. I think it is not bad idea to rimraf this dir too in clean task. What do you think?

Any thoughts?

sharikovvladislav commented:
Oops. Not it was real error. Any thoughts why? I will debug it today a not later.

sharikovvladislav commented on packages/stryker-api/src/config/ConfigWriterFactory.ts:
```diff
@@ -1,12 +1,12 @@
    * Represents a Factory for ConfigWriters.
    */
-  class ConfigWriterFactory extends Factory<void, ConfigWriter> {
+  class ConfigWriterFactory extends Factory<void, Configurator> {
```
Fixed

sharikovvladislav commented on packages/stryker-api/src/config/ConfigWriterFactory.ts:
```diff
@@ -1,12 +1,12 @@
 import {Factory} from '../../core';
```
Fixed

simondel commented:
@sharikovvladislav It looks good! Thanks for the hard work :) I don't mind if the `.nyc_output` is not cleaned with every build, but it is better to clean it just to be sure.

sharikovvladislav commented:
@simondel maybe. I had a bit another case. It contains a lot of autgenerated reports with class/file names. I had case where I should refactor class/file names. `.nyc_output` had produced some noise. When I try to find something a lot of results were in autogenerated folders. In that case `npm run clean` is not bad idea :)

I bet you will have same tasks in the future and I hope it will be easier for others in future.
